File: app/xsquad/path.py
from operator import itemgetter
from Queue import PriorityQueue
from math import sqrt

from los import key2point, point2key


# stolen from los.pyx... should not be duplicated

EMPTY1 = "."
EMPTY = EMPTY1 * 6

def west(p):
    return p[0] != EMPTY1

def east(p):
    return p[1] != EMPTY1

def south(p):
    return p[2] != EMPTY1

def north(p):
    return p[3] != EMPTY1

def up(p):
    return p[4] != EMPTY1

def down(p):
    return p[5] != EMPTY1


def position_delta(key1, key2):
    a = key2point(key1)
    b = key2point(key2)
    return (b[0] - a[0], b[1] - a[1], b[2] - a[2])


def get_neighbor_key(key, dx=0, dy=0, dz=0):
    x, y, z = key2point(key)
    return point2key((x+dx, y+dy, z+dz))


def is_passable(sides, dx, dy):
    print sides
    if dx == -1:
        if west(sides):
            return False
        if dy == -1:
            return not south(sides)
        if dy == 1:
            return not north(sides)
        return True
    if dx == 1:
        if east(sides):
            return False
        if dy == -1:
            return not south(sides)
        if dy == 1:
            return not north(sides)
        return True
    if dy == -1:
        return not south(sides)
    if dy == 1:
        return not north(sides)


def get_neighbors(key, world):
    """
    This is where we figure out which of the neighboring positions
    can be moved to. We count directions in the horizontal plane,
    up/down is treated as special cases.

    A wall in one direction means that we also can't move to the
    diagonals touching that direction.  This means it's not
    allowed to cut corners or move diagonally through doors.
    (It also simplifies the calculation a lot :)
    """
    sides = world[key]
    neighbors = []
    for dx in (-1, 0, 1):
        for dy in (-1, 0, 1):
            if dx != 0 or dy != 0:
                neigh_key = get_neighbor_key(key, dx, dy)
                neigh_sides = world.get(neigh_key, EMPTY)
                # TODO: check if block is within level borders
                if (is_passable(sides, dx, dy) and
                        is_passable(neigh_sides, -dx, -dy)):
                    if down(neigh_sides):
                        neighbors.append(
                            {"key": neigh_key,
                             "cost": 3 if abs(dx) == abs(dy) else 2})
                    else:
                        below = get_neighbor_key(neigh_key, 0, 0, -1)
                        if below in world and down(world[below]):
                            neighbors.append({"key": below, "cost": 5})
                elif abs(dx) != abs(dy):
                    above_neigh = get_neighbor_key(neigh_key, 0, 0, 1)
                    above = get_neighbor_key(key, 0, 0, 1)
                    if above_neigh in world and down(world[above_neigh]) and is_passable(world.get(above, EMPTY), dx, dy):
                        neighbors.append({"key": above_neigh, "cost": 7})
    return sorted(neighbors, key=itemgetter("cost"))


def make_path_graph(world):
    connections = {}
    for key in world:
        neighbors = get_neighbors(key, world)
        if neighbors:
            connections[key] = neighbors
    return connections


# A* path finding (not really used yet, untested)

def heuristic(key1, key2, weight=5):
    x1, y1, z1 = key2point(key1)
    x2, y2, z2 = key2point(key2)
    return weight * sqrt((x2-x1)**2 + (y2-y2)**2 + (z2-z1)**2)


def find_path(startkey, goalkey, world):

    frontier = PriorityQueue()
    frontier.put({"key": startkey, "cost": 0}, 0)
    came_from = {startkey: None}
    cost_so_far = {startkey: 0}
    step = 0

    while not frontier.empty() and step < 100000:
        step += 1  # just safety
        current = frontier.get()
        if current["key"] == goalkey:
            break

        for neighbor in get_neighbors(current["key"], world):
            new_cost = cost_so_far[current["key"]] + neighbor["cost"]
            if new_cost < cost_so_far.get(neighbor["key"], 10000000):
                cost_so_far[neighbor["key"]] = new_cost
                neighbor["cost"] += heuristic(neighbor["key"], goalkey)
                frontier.put(neighbor)
                came_from[neighbor["key"]] = current["key"]

    if current["key"] != goalkey:
        return None

    key = goalkey
    path = []
    step = 0
    while key and step < 1000:
        key = came_from[key]
        path.insert(0, key)
    return path

File: app/static/js/path.js
PathGraph = (function () {

    function _Path(graph) {
        this.graph = graph;
    }

    /* Note: all the helper functions are not used anymore since
     we now get the pathfindig graph from the server.
     Kept for sentimental reasons. */

    function positionDelta(key1, key2) {
        var a = key2point(key1), b = key2point(key2);
        return {dx: b[0] - a[0], dy: b[1] - a[1], dz: b[2] - a[2]};

    }

    function getNeighborKey(key, dx, dy, dz) {
        var z =  key >> 16;
        var y = (key >> 8) - z * 256;
        var x = key - y*256 - z*256*256;
        return x + dx + ((y + dy) << 8) + ((z + dz) << 16);
    }


    function isPassable(sides, dx, dy) {
        if (!sides)
            return true;
        if (dx == -1) {
            if (sides[0] != ".")
                return false;
            if (dy == -1) return (sides[2] == ".");
            if (dy == 1) return (sides[3] == ".");
            return true;
        }
        if (dx == 1) {
            if (dy == -1) return (sides[2] == ".");
            if (dy == 1) return (sides[3] == ".");
            return sides[1] == ".";
        }
        if (dy == -1) {
            return sides[2] == ".";
        }
        if (dy == 1)
            return sides[3] == ".";
    }

    function getNeighbors(key, world, from) {
        /*
         This is where we figure out which of the neighboring positions
         can be moved to. We count directions in the horizontal plane,
         up/down is treated as special cases.

         A wall in one direction means that we also can't move to the
         diagonals touching that direction.  This means it's not
         allowed to cut corners or move diagonally through doors.
         (It also simplifies the calculation a lot :)
         */
        var sides = world[key], neighborKey, neighbors = [];
        for (var dx = -1; dx <= 1; dx += 1) {
            for (var dy = -1; dy <= 1; dy += 1) {
                neighborKey = getNeighborKey(key, dx, dy, 0);
                if (!(dx === 0 && dy === 0)) {
                    if (isPassable(world[key], dx, dy) && isPassable(world[neighborKey], -dx, -dy)) {
                        if ((world[neighborKey] >> 10) & 1) {  // is there a floor here?
                            neighbors.push({key: neighborKey,
                                            cost: Math.abs(dx) === Math.abs(dy)? 3 : 2});
                        } else  {
                            // no floor! well, is there something below?
                            // max fall height is 1 unit
                            var below = getNeighborKey(neighborKey, 0, 0, -1);
                            if (R.has(below, world) && (world[below] >> 10) & 1) {
                                console.log("below", key2point(below));
                                neighbors.push({key: below, cost: 5});
                            }
                        }
                    } else if (Math.abs(dx) !== Math.abs(dy)) {
                        // wall eh? Let's check if it can be scaled
                        // maximum climb is 1 unit
                        var above = getNeighborKey(key, dx, dy, 1);
                        if (R.has(above, world) && (world[above] >> 10) & 1) {
                            console.log("above", key2point(above));
                            neighbors.push({key: above, cost: 7});
                        }
                    }
                }
            }
        }
        return neighbors;
    }

    function heuristic(key1, key2) {
        // Manhattan distance on a square grid
        var p1 = key2point(key1), p2 = key2point(key2);
        return 2 * Math.sqrt(Math.pow(p1[0] - p2[0], 2) + Math.pow(p1[1] - p2[1], 2) + Math.pow(p1[2] - p2[2], 2));
    }

    function findPath (start, goal) {

        console.log("findPath", key2point(start), key2point(goal));

        var frontier = new PriorityQueue(
            {comparator: function (a, b) {return a.cost - b.cost;}});
        frontier.queue({key: start, cost: 0});
        var cameFrom = {};
        cameFrom[start] = -1;
        var costSoFar = {};
        costSoFar[start] = 0;
        var current, last, neighbors, totalCost;
        var step = 0;

        while (frontier.length > 0 && step++ < 1000000) {
            current = frontier.dequeue();
            if (current.key == goal) {
                totalCost = costSoFar[current.key];
                break;
            }
            neighbors = this.graph[current.key];

            var newCost, next;
            neighbors.forEach(function (next, i) {
                newCost = costSoFar[current.key] + next.cost;
                if (!R.has(next.key, costSoFar) || (newCost < costSoFar[next.key])) {
                    costSoFar[next.key] = newCost;
                    frontier.queue({key: next.key,
                                    cost: next.cost + heuristic(next.key, goal)});
                    cameFrom[next.key] = current.key;
                }
            });
            last = current;
        }

        // check if we were able to find a suitable path
        if (current.key != goal) {
            return;
        }

        current = goal;
        var path = [current];
        step = 0;
        while (current != -1 && step++ < 10000) {
            current = cameFrom[current];
            path.unshift(current);
        }

        if (path.length > 0 && totalCost)
            return {points: path.slice(1), cost: totalCost};
    }

    function makePathGraph (world) {
        var connections = {};
        Object.keys(world).forEach(function (key) {
            var neighbors = getNeighbors(key, world);
            console.log("neighbors", key, neighbors);
            if (neighbors.length > 0)
                connections[key] = neighbors;
        });
        console.log("pathGrapoh", connections);
        return connections;
    }

    _Path.prototype.findPath = R.memoize(findPath);

    //return {findPath: findPath, makePathGraph: makePathGraph};
    return _Path;

})();
